frame/fast-unstake/src/lib.rs

```diff
@@ -0,0 +1,558 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A pallet that's designed to ONLY do:
+//!
+//! If a nominator is not exposed at all in any `ErasStakers` (i.e. "has not actively backed any
+//! validators in the last `BondingDuration` days"), then they can register themselves in this
+//! pallet, unstake faster than needing to wait an entire bonding duration, and potentially move
+//! into a nomination pool.
+//!
+//! Appearing in the exposure of a validator means being exposed equal to that validator from the
+//! point of view of the staking system. This usually means earning rewards with the validator, and
+//! also being at the risk of slashing with the validator. This is equivalent to the "Active
+//! Nominator" role explained in the
+//! [February Staking Update](https://polkadot.network/blog/staking-update-february-2022/).
+//!
+//! This pallet works of the basis of `on_idle`, meaning that it provides no guarantee about when it
+//! will succeed, if at all. Moreover, the queue implementation is unordered. In case of congestion,
+//! not FIFO ordering is provided.
+//!
+//! Stakers who are certain about NOT being exposed can register themselves with
+//! [`Call::register_fast_unstake`]. This will chill, and fully unbond the staker, and place them in
+//! the queue to be checked.
+//!
+//! Once queued, but not being actively processed, stakers can withdraw their request via
+//! [`Call::deregister`].
+//!
+//! Once queued, a staker wishing to unbond can perform to further action in pallet-staking. This is
+//! to prevent them from accidentally exposing themselves behind a validator etc.
+//!
+//! Once processed, if successful, no additional fees for the checking process is taken, and the
+//! staker is instantly unbonded. Optionally, if the have asked to join a pool, their *entire* stake
+//! is joined into their pool of choice.
+//!
+//! If unsuccessful, meaning that the staker was exposed sometime in the last `BondingDuration` eras
+//! they will end up being slashed for the amount of wasted work they have inflicted on the chian.
```

Comment:
This is a bit heavy-handed for stakers who did not know beforehand whether they've been exposed. However, we could view this as a UI issue by disabling the ability to `register_fast_unstake` when the frontend detects that the staker had been exposed in the last `BondingDuration`.